crates/ruff_linter/src/rules/ruff/rules/suppression_comment_visitor.rs

```diff
@@ -0,0 +1,272 @@
+    }
+    fn visit_body(&mut self, body: &'ast [ruff_python_ast::Stmt]) {
+        match body {
+            [] => {
+                // no-op
+            }
+            [only] => {
+                self.visit_stmt(only);
+            }
+            [first, .., last] => {
+                if self.can_skip(last.end()) {
+                    // Skip traversing the body when there's no comment between the first and last statement.
+                    // It is still necessary to visit the first statement to process all comments between
+                    // the previous node and the first statement.
+                    self.visit_stmt(first);
+                    self.preceding_node = Some(last.into());
+                } else {
+                    preorder::walk_body(self, body);
+                }
+            }
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub(super) struct SuppressionCommentData<'src> {
+    /// If `enclosing` is `None`, this comment is top-level
+    pub(super) enclosing: Option<AnyNodeRef<'src>>,
+    pub(super) preceding: Option<AnyNodeRef<'src>>,
+    pub(super) following: Option<AnyNodeRef<'src>>,
+
+    pub(super) line_position: CommentLinePosition,
+    pub(super) kind: SuppressionKind,
+    pub(super) range: TextRange,
+}
+
+pub(super) trait CaptureSuppressionComment<'src> {
+    /// This is the entrypoint for the capturer to analyze the next comment.
+    fn capture(&mut self, comment: SuppressionCommentData<'src>);
+}
+
+/// Determine the indentation level of an own-line comment, defined as the minimum indentation of
+/// all comments between the preceding node and the comment, including the comment itself. In
+/// other words, we don't allow successive comments to ident _further_ than any preceding comments.
+///
+/// For example, given:
+/// ```python
+/// if True:
+///     pass
+///     # comment
+/// ```
+///
+/// The indentation would be 4, as the comment is indented by 4 spaces.
+///
+/// Given:
+/// ```python
+/// if True:
+///     pass
+/// # comment
+/// else:
+///     pass
+/// ```
+///
+/// The indentation would be 0, as the comment is not indented at all.
+///
+/// Given:
+/// ```python
+/// if True:
+///     pass
+///     # comment
+///         # comment
+/// ```
+///
+/// Both comments would be marked as indented at 4 spaces, as the indentation of the first comment
+/// is used for the second comment.
+///
+/// This logic avoids pathological cases like:
+/// ```python
+/// try:
+///     if True:
+///         if True:
+///             pass
+///
+///         # a
+///             # b
+///         # c
+/// except Exception:
+///     pass
+/// ```
+///
+/// If we don't use the minimum indentation of any preceding comments, we would mark `# b` as
+/// indented to the same depth as `pass`, which could in turn lead to us treating it as a trailing
+/// comment of `pass`, despite there being a comment between them that "resets" the indentation.
+pub(super) fn own_line_comment_indentation(
```

Comment:
We can't move it to the trivia crate, unfortunately, because of a circular dependency with `ruff_python_ast`. `invalid_formatter_suppression_comments` could work, though.